Documentation/ceph-best-practices.md

```diff
@@ -0,0 +1,516 @@
+To implement this in Rook, we ensure that Rook will only schedule MONs on nodes with MON SSDs. We
+therefore have a required Affinity for those nodes. We can accomplish this by applying a
+`ceph-mon-ssd=true` label to nodes with SSDs for Ceph MONs. Note that the `mon` section’s
+`nodeAffinity` takes precedence over and overwrites the `all` section’s `nodeAffinity` for Ceph MON
+daemons. Make sure that you re-specify the rules from the `all` section to ensure Ceph MONs also
+continue to maintain Affinity only for storage nodes as we configured above.
+
+
+**`cluster.yaml`:**
+```yaml
+placement:
+  mon:
+    nodeAffinity:​
+      requiredDuringSchedulingIgnoredDuringExecution:​
+        nodeSelectorTerms:​
+          # respecify the 'all' nodeAffinity rules here for Ceph MON daemons
+          - matchExpressions:​
+              - key: role​
+                operator: In​
+                values:​
+                  - storage-node
+          # add affinity for nodes with MON SSDs
+          - matchExpressions:​
+              - key: ceph-mon-ssd​
+                operator: In​
+                values:​
+                  - "true"
+```
+
+We want to schedule MONs so they are spread across failure domains whenever possible. We will
+accomplish this by applying Anti-affinity between MON pods. Rook labels all MON pods
+`app=rook-ceph-mon`, and that is what will be used to spread the monitors apart. There is one rule
+for rooms, and one for racks if a room is down. We want to ensure a higher `weight` is given to
+riskier co-location scenarios.
+
+We do not recommend running MONs on the same node unless absolutely necessary. Rook automatically
+applies an Anti-affinity with medium-level `weight`. However, this might not be appropriate for all
+scenarios. For our scenario, we only want node-level co-location in the worst of failure scenarios,
+so we want to apply the highest possible `weight` of Anti-affinity for nodes.
+
+**`cluster.yaml`:**
+```yaml
+placement:
+  mon:
+    # ... nodeAffinity from above ...
+    podAntiAffinity:​
+      preferredDuringSchedulingIgnoredDuringExecution:​
+        - weight: 80​
+          podAffinityTerm:​
+            labelSelector:​
+              matchLabels:​
+                app: rook-ceph-mon​
+            topologyKey: topology.rook.io/room​
+        - weight: 90​
+          podAffinityTerm:​
+            labelSelector:​
+              matchLabels:​
+                app: rook-ceph-mon​
+            topologyKey: topology.rook.io/rack​
+        - weight: 100​
+          podAffinityTerm:​
+            labelSelector:​
+              matchLabels:​
+                app: rook-ceph-mon​
+            topologyKey: kubernetes.io/hostname​
+```
+
+> **NOTE:** If host networking is enabled, you cannot co-locate MONs because the ports will collide
+> on nodes. To enforce this, if host networking is enabled, Rook will automatically set a
+> `requiredDuringSchedulingIgnoredDuringExecution` Pod Anti-affinity rule.
+
+### Ceph OSDs
```

Comment:
This section doesn't talk about storageClassDeviceSets, which is critical for more dynamic environments using PVCs.

Reply:
But since the focus is on the specific advanced walkthrough, no need to discuss device sets in this doc.